### ISSUE TITLE
Unable to build Chapel 1.25.0 under Gentoo Linux with CHPL_LLVM=system

### ISSUE THREAD

**Issue by preney** [chapel-lang/chapel#18589]
### Summary of Problem
When using these steps to compile Chapel v1.25.0 everything compiles correctly:

        wget https://github.com/chapel-lang/chapel/releases/download/1.25.0/chapel-1.25.0.tar.gz
        tar xvf chapel-1.25.0.tar.gz
        cd chapel-1.25.0
        export CHPL_LLVM=bundled
        source util/setchplenv.bash
        make

however when using these steps to use the system-installed LLVM/Clang:

        wget https://github.com/chapel-lang/chapel/releases/download/1.25.0/chapel-1.25.0.tar.gz
        tar xvf chapel-1.25.0.tar.gz
        cd chapel-1.25.0
        export CHPL_LLVM=system
        export CHPL_LLVM_CONFIG=/usr/lib/llvm/11/bin/llvm-config
        export CHPL_HOST_COMPILER=llvm
        export CHPL_TARGET_COMPILER=llvm
        source util/setchplenv.bash
        make

compilation fails with these errors:

        ld.lld: error: unable to find library -lclangFrontend
        ld.lld: error: unable to find library -lclangSerialization
        ld.lld: error: unable to find library -lclangDriver
        ld.lld: error: unable to find library -lclangCodeGen
        ld.lld: error: unable to find library -lclangParse
        ld.lld: error: unable to find library -lclangSema
        ld.lld: error: unable to find library -lclangAnalysis
        ld.lld: error: unable to find library -lclangEdit
        ld.lld: error: unable to find library -lclangASTMatchers
        ld.lld: error: unable to find library -lclangAST
        ld.lld: error: unable to find library -lclangLex
        ld.lld: error: unable to find library -lclangBasic
        clang-11: error: linker command failed with exit code 1 (use -v to see invocation)

Since these libclang libraries are not on my system I submitted a bug report to Gentoo Linux here:

* https://bugs.gentoo.org/818403

and the response was to use `-lclang-cpp`. (Per https://archives.gentoo.org/gentoo-dev/message/6d3cf88d858fcf3fbb11818ce5d6ea42 Gentoo as of Clang v10 was no longer using BUILD_SHARED_LIBS=ON when building Clang and instead uses the 'dylib' model).

I grep'd the chapel source code for these libraries and they are mentioned in `third-party/llvm/Makefile.include-system` so I changed the file so that:

        LLVM_CLANG_LIBS=-lclang-cpp

was set and I still get linker errors. After experimentation, starting with:

        export CHPL_LLVM=system
        export CHPL_LLVM_CONFIG=/usr/lib/llvm/11/bin/llvm-config
        export CHPL_HOST_COMPILER=llvm
        export CHPL_TARGET_COMPILER=llvm
        export CHPL_LLVM_DYNAMIC=1
        export CHPL_LIB_PIC=pic

and then editing `third-party/llvm/Makefile.include-system` so the following are set:

        LLVM_CLANG_LIBS=-lclang-cpp
        # Add -fPIC...
        LLVM_CXXFLAGS=-fPIC $(LLVM_CONFIG_CXXFLAGS) $(LLVM_MY_CXXFLAGS) -DHAVE_LLVM
        LLVM_CFLAGS=-fPIC $(LLVM_CONFIG_CFLAGS) -DHAVE_LLVM
        # Add -dynamic...
        LLVM_LIBS=-dynamic -shared -L$(LLVM_CONFIG_LIB_DIR) $(LLVM_CLANG_LIBS) $(LLVM_LLVM_LIBS)

and then running `source util/setchplenv.bash` followed by `make` yields a successful compile --but when `chpl` is run (including with no arguments) it immediately segfaults, e.g.,

        $ chpl
        Segmentation fault (core dumped)

What behavior did you observe when encountering this issue?
* See above,

What behavior did you expect to observe?
* A successful compilation using the system LLVM/Clang.

Is this a blocking issue with no known work-arounds?
* A work-around is to use `CHPL_LLVM=bundle`.

There are no instructions given in Chapel concerning this, but, my system is running LLVM/Clang is v11.1.0 --not v11.0.1 as stated in `third-party/llvm/README`. The Chapel docs say LLVM/Clang v11 --which I am running.

I suppose it is possible that when system LLVM is being used, the Chapel build might be incorrectly using source code from the bundled code --not the system (e.g., via header files). If so then this could be causing issues. Since the organization of code in `third-party/llvm` doesn't appear to match pristine LLVM/Clang code, I did not replace any LLVM/Clang v11.0.1 code with the same from v11.1.0. (Additionally, there is no documentation concerning how to graft such in or how to remove the all bundled LLVM/Clang 11.0.1 code (since it shouldn't be used at all) from `third-party/llvm` --so I left such alone.)

In summary, the problem appears to be that the build code for Chapel assume the ability to statically link to LLVM/Clang libraries when using a system-installed LLVM/Clang. On its own `CHPL_LLVM_DYNAMIC`  doesn't fix any issues: it fails to link (even if the `-lclang-cpp` edits are made as detailed above). One has to use `-fPIC` and `-dynamic -shared` as discussed above to get things to compile and link (but it segfaults when run). How might this be fixed so Chapel can be built using the system's LLVM/Clang v11? Thanks.

(ASIDE: All Chapel prerequisites are installed in the system.)

**Comment by preney** [comment]
Hi @mppf ! My apologies for the delay in response... but with this post you will infer, I've been busy creating an straight-forward solution so you could have an installation of Gentoo Linux to work with ... and the same could be used to have other Linux distributions (including multiple versions) installed for Chapel testing purposes as well + such can all be scripted. :-)

I apologize as this post is long (and this isn't really a normal issue/bug post at all). I have created a solution involving a Bash shell script and a Singularity (i.e., https://singularity.hpcng.org/ https://github.com/hpcng/singularity or commercially https://sylabs.io/singularity/ ) definition file (which is a "script" that creates a Singularity container). Singularity is (Linux) container technology that is ideal for high-performance computing (HPC) environments since it is also secure and using a container can completely occur in user space without any daemons, etc. (unlike Docker). (ASIDE: My day job is involves supporting Canadian academic research via SHARCNET/Compute Canada (e.g., https://www.sharcnet.ca and https://www.computecanada.ca/ ).)

While some readers of this might know about and use Singularity others might not, so in this post I provide hopefully sufficient information to help with the installation of Singularity. Before discussing such, however, it is important to answer this question, "Why should one even consider using Singularity?" Answer: Reproducibility. Every system is an amalgam of all kinds of software, compiled with all kinds of configuration settings, etc. and while we hope some low probability random thing isn't a contributing factor --it might be. LLVM/Clang is a complicated piece of software with all kinds of options that takes approximately 30 minutes to compile on a 16-core AMD Ryzen Threadripper PRO 3955WX system. While we hope various settings that LLVM/Clang can be compiled with aren't an issue, it is unrealistic to say those will never be significant problems. Having reproducible containers will allow us to produce exactly the same containers + discuss/explore/resolve those issues in a sane manner.

ASIDE: Gentoo Linux builds everything from source. That said, an initial system starts of will a `stage3` (binary) tarball --which the Singularity .def file mentioned below downloads and puts into the container. (One can tell Gentoo's package manager to rebuild such if one wants --I did not do this in the files mentioned below.) Because Gentoo compiles things from source and is all about configurability, reproducibility is important to have in this instance with the end goal of hopefully being able to make Chapel's build system more robust to underlying system configurations. The latter is important to take advantage of things done on systems. For example, HPC compute cluster systems will have all kinds of tools custom built to take advantage of installed hardware, fabrics, etc. to better performance. Software tools that can make use of such typically get those advantages "for free" if they can use those already installed tools.

To enable reproducibility and discussions (e.g., to deal with this Github issue), I created two files:
* demo.sh
* singularity-gentoo-amd64-chapel-basis.def
(Links to these files are provided later in this post.)

The demo.sh file is a Bash script that, when run, completely automates the following:
* building a pristine default options-set Gentoo Linux (Singularity) container from scratch (ASIDE: Know that Gentoo Linux builds everything from source and its "thing" is all about (re)configuribility.)
* it copies the Gentoo Linux container (once per case instance) to a case# for testing purposes --and then installs LLVM/Clang using Gentoo's package manager (with different settings) + Chapel v1.25.0.

The .def file is a script that Singularity can use to build an container of Gentoo Linux (plus Chapel prerequisite software installed).

I wrote such to:
* make it easy to make adjustments/corrections, explore, etc. each case individually including exploring fixes, etc.,
* to have the same results for anyone also using the same ./demo.sh script and case inputs, and,
* to easily add more cases + share such.

Assuming one has an installed Linux system with typical shell commands installed along with `sudo` and Singularity installed, then running `./demo.sh` will do the following:
* In the current directory a Singularity "sandbox" (think "chroot"/filesystem directory) image will be create with Gentoo Linux installed in it + the gentoo.log capturing everything output during its creation/installation.
* In the current directory a number (i.e., six right now) of test cases in case##.dir directories with case##.log files of Singularity containers that started out as copies of gentoo.dir/ and had Clang v11 installed (using various settings) + attempts to install Chapel under such.

After `./demo.sh` finishes running, one can then explore the images. For example, to explore case02.dir, one would first run the following:
* `singularity shell -C -w case02.dir`
(ASIDE: The Chapel build attempt is in `/testing/chapel` and the rest of the image is Gentoo Linux.)

To give you an idea of resources needed: running `demo.sh` on a 16-core AMD Ryzen Threadripper PRO 3955WX takes approximately 30 minutes or so for each case after building the base Gentoo Linux container used by all cases. (For some reason Gentoo also pulls in Clang v12 which is also built when installing Clang v11.) Each case's log file is about 20MB and each case directory is about 3.3GB. (gentoo.dir is about 1.9GB and gentoo.log about 6.8MB.) The binary architecture targets 64-bit Linux (I didn't set a specific architecture.) The script does set the environment variable MAKEOPTS to take advantage of parallel building which is useful since building LLVM/clang is not fast --but you can edit/delete such if you want.

You can obtain a copy of the files needed to do this from the Gentoo issue here (let me know where you might want such posted here as it doesn't seem like I can attach files to a post in Github):
* demo.sh is linked to from Comment 7 here: https://bugs.gentoo.org/818403#c7
* singularity-gentoo-amd64-chapel-basis.def is linked to from Comment 6 here: https://bugs.gentoo.org/818403#c6

But before you can run such, you need to have `sudo`, the "standard" Linux command line tools, `tar`, `wget`, and Singularity all installed. Many Linux distributions don't have Singularity available as a package (some might appear to but often it is a completely different program) so one may have to build Singularity from source --which is not hard to do.

If Singularity is not available with your package manager or on a Linux system, you will need to install it from source which is detailed here: https://singularity.hpcng.org/user-docs/3.8/quick_start.html . Ensure the prerequisite tools mentioned are installed, then install Go, and then it is very easy to install Singularity. (You might want to also run `sudo apt install build-essential` (Debian/Ubuntu) or `sudo yum groupinstall 'Development Tools'` (CentOS, Fedora, other RedHat distributions) to install common code building compilers and utilities )

After the software prerequisites and Singularity have been installed, you should be able to successfully run `./demo.sh`. The code in both files is straight-forward. If you've any questions, please ask.) After running such, the results are:

* `case01` and `case04` properly build and allow the Chapel compiler to run. (Yay!)
* The remaining cases (2, 3, 5, and 6) fail.

Setting aside how Chapel was built (all of these cases did exactly the same commands + there was one (1) edit to one file in the downloaded Chapel v1.25.0 tarball (see later in this post), the differences between these cases is only how the LLVM/Clang compiler was compiled involving these three Gentoo "USE" flags:
* `default-compiler-rt             : Use compiler-rt instead of libgcc as the default rtlib for clang`
* `default-libcxx                  : Use libc++ instead of libstdc++ as the default stdlib for clang`
* `default-lld                     : Use lld as the default linker for clang`
i.e.,
* `case01` used the default, i.e., none of these USE flags were set.
* `case04` built clang with the `default-lld` USE flag set.
* `case02` built clang with the `default-compiler-rt` USE flag set.
* `case03` built clang with the `default-libcxx` USE flag set,
* `case05` built clang with the `default-compiler-rt` and `default-lld` USE flags set.
* `case06` built clang with the `default-compiler-rt`, `default-libcxx`, and `default-lld` USE flags set.

(Of these `case06` is not a Chapel issue: it failed because of a package manager "block" refusing to even install Clang --so this is not a Chapel issue and can be ignored here.)

From this (without further explorations, etc.) it appears that Chapel will build and run using these settings:
```
export CHPL_LLVM=system
export CHPL_LLVM_CONFIG=/usr/lib/llvm/11/bin/llvm-config
export CHPL_HOST_COMPILER=llvm
export CHPL_HOST_CC=clang-11
export CHPL_HOST_CXX=clang++-11
export CHPL_TARGET_COMPILER=llvm
export CHPL_TARGET_CC=clang-11
export CHPL_TARGET_CXX=clang++-11
sed -i -e 's/LLVM_CLANG_LIBS=.*/LLVM_CLANG_LIBS=-lclang-cpp/' third-party/llvm/Makefile.include-system
source util/setchplenv.bash
make -j$(nproc)
```
Notice the change to only have `-lclang-cpp` in the LLVM Makefile as previous mentioned in this issue. I also explicitly set the compilers since Clang v12 was also installed. (This might be overkill for Chapel --but for now reduces a variable.) The result of these cases appear to demonstrate that only when the Gentoo `default-compiler-rt` and `default-libcxx` USE flags are not set, Chapel builds successfully. Assuming this isn't a random/weird issue arising from Gentoo Linux, this implies there may be issues with Chapel when it is not using libgcc and libstdc++ under Linux where GCC is the main compiler. As I have been busy creating a solution that can be easily adopted and used by you and the Chapel team, I've not explored this further.

Should you have issues installing Singularity or don't have `sudo` access on a machine with Singularity kindly let me know. As this post is long enough so I will end it here for now --but please do let me know of any issues with the above, etc.:-)

Paul

P.S. I am open to using audio/video conferencing should you want to audio/video conference to discuss any of the above, etc. Let me know if you would like to do such. (My time zone is Eastern Time (North America).)
